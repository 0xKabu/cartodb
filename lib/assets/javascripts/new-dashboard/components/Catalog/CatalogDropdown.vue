<template>
  <div 
    class="catalogDropdown"
    @keydown.down.prevent="onKeydownDown"
    @keydown.up.prevent="onKeydownUp"
    :class="{'is-disabled': isDisabled, 'is-open': isOpen}">
    <span class="title is-caption catalogDropdown__label">{{ title }}</span>
    <div class="catalogDropdown__container">
      <input type="text"
        ref="catalogDropdownInput"
        class="text is-caption catalogDropdown__input"
        :placeholder="placeholder"
        v-model="searchFilter"
        @focus="onInputFocus"
        @click="openDropdown"
        @keyup.enter="onKeyEnter"
        :disabled="isDisabled">
        <button v-if="searchFilter" class="catalogDropdown__close" @click="resetInput"><img src="../../assets/icons/common/dropdown-close.svg" width="16" height="20" /></button>
      <ul class="catalogDropdown__list" 
        :class="{'is-open': isOpen, 'is-height-limited': limitHeight}"
        @mouseleave="resetActiveOption">
        <li
          v-for="(option, index) in filteredOptions" :key="option"  
          @click="selectOption(option)"
          class="catalogDropdown__option text is-caption"
          :class="{'catalogDropdown__option--active': activeOptionIndex === index + 1}"
          @mouseover="updateActiveOption(index + 1)">
          {{ option }}
        </li>
      </ul>
      <div class="catalogDropdown__extra"><span class="catalogDropdown__extra--text text is-small">Your country is not listed <a href="">Contact Us</a></span></div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'CatalogDropdown',
  props: {
    title: String,
    placeholder: String,
    options: {
      type: Array,
      default: []
    },
    open: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    limitHeight: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      selected: {},
      isInputFocused: false,
      activeOptionIndex: -1,
      searchFilter: '',
      isOpen: this.open,
      isDisabled: this.disabled
    }
  },
  computed: {
    filteredOptions() {
      const filtered = [];
      const regOption = new RegExp(this.searchFilter, 'ig');
      for (const option of this.options) {
        if (this.searchFilter.length < 1 || option.match(regOption)){
          filtered.push(option);
        }
      }
      return filtered.length > 0 ? filtered : this.options ;
    }
  },
  methods: {
    openDropdown () {
      this.isOpen = true;
    },
    closeDropdown () {
      this.isOpen = false;
    },
    onInputFocus () {
      this.isInputFocused = true;
    },
    onKeydownDown () {
      if (this.activeOptionIndex < this.options.length) {
        this.activeOptionIndex++;
      }
    },
    onKeydownUp () {
      if (this.activeOptionIndex > 0) {
        this.activeOptionIndex--;
      }
    },
    onKeyEnter () {
      const optionSelected = this.$el.querySelector('.catalogDropdown__option--active');
      if (optionSelected) {
        optionSelected.click();
      }
    },
    updateActiveOption (index) {
      this.activeOptionIndex = index;
    },
    resetActiveOption () {
      this.activeOptionIndex = -1;
    },
    selectOption (option) {
      this.selected = option;
      this.isOpen = false;
      this.searchFilter = this.selected;
      this.$emit('selected', this.selected);
    },
    resetInput () {
      this.searchFilter = '';
      this.selected = {};
      this.$refs.catalogDropdownInput.value = '';
      this.$emit('reset');
    }
  },
  watch: {
    disabled() {
      this.isDisabled = this.disabled;
    },
    open() {
      this.isOpen = this.open;
    }
  }
};
</script>

<style scoped lang="scss">
@import 'new-dashboard/styles/variables';

.catalogDropdown {
  display: flex;
  position: absolute;
  z-index: 2;
  flex-direction: column;
  width: 100%;

  &__label {
    margin-bottom: 16px;
  }

  &__container {
    position: absolute;
    top: 48px;
    width: 100%;
  }

  &__input {
    width: 100%;
    height: 56px;
    padding: 16px 24px;
    border: 1px solid $neutral--300;
    border-radius: 4px;
  }

  &__close {
    position: absolute;
    top: 18px;
    right: 24px;
  }

  &.is-disabled {
    opacity: 0.24;

    .catalogDropdown__close {
      display: none;
    }
  }

  &.is-open {
    .catalogDropdown__input {
      border-top: transparent;
      border-right: transparent;
      border-left: transparent;
      border-bottom-right-radius: 0;
      border-bottom-left-radius: 0;
    }

    .catalogDropdown__list {
      border: transparent;
    }

    .catalogDropdown__container {
      border: 2px solid $primary-color;
      border-radius: 4px;
    }

    .catalogDropdown__extra {
      visibility: visible;
      opacity: 1;
      pointer-events: auto;
    }
  }
}

.catalogDropdown__list {
  visibility: hidden;
  border: 1px solid $softblue;
  opacity: 0;
  background-color: $white;
  pointer-events: none;

  &.is-open {
    visibility: visible;
    opacity: 1;
    pointer-events: auto;
  }

  &.is-height-limited {
    max-height: 196px;
    overflow-y: scroll;
  }
}

.catalogDropdown__option {
  display: block;
  position: relative;
  width: 100%;
  padding: 12px 16px 12px 36px;
  overflow: hidden;
  color: $primary-color;
  text-decoration: none;
  text-overflow: ellipsis;
  white-space: nowrap;
  cursor: pointer;

  &:not(:last-child) {
    border-bottom: 1px solid $softblue;
  }

  &--active {
    background-color: rgba($primary-color, 0.05);
    text-decoration: underline;
  }
}

.catalogDropdown__extra {
  visibility: hidden;
  padding: 16px 24px;
  border-top: 1px solid $softblue;
  opacity: 0;
  background-color: $white;
  pointer-events: none;

  &--text {
    color: $neutral--600;
  }
}
</style>
