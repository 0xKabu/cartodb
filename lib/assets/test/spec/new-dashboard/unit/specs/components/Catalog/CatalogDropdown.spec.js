import { shallowMount } from '@vue/test-utils';
import CatalogDropdown from 'new-dashboard/components/Catalog/CatalogDropdown';
import PerfectScrollbar from 'perfect-scrollbar';

const $t = key => key;

describe('CatalogDropdown.vue', () => {
  function createCatalogDropdown (overridenComponentOptions) {
    return shallowMount(CatalogDropdown, {
      mocks: {
        $t
      },
      propsData: {
        title: 'Title',
        placeholder: 'Placeholder',
        options: [],
        open: true,
        disabled: false,
        limitHeight: false
      },
      slots: {
        extra: 'Extra Content'
      },
      ...overridenComponentOptions
    });
  }

  beforeAll(() => {
    spyOn(PerfectScrollbar, 'initialize').and.returnValue('initialized');
    spyOn(PerfectScrollbar, 'update').and.returnValue('updated');
    spyOn(PerfectScrollbar, 'destroy').and.returnValue('destroyed');
  });

  it('should render correct contents', () => {
    const catalogDropdown = createCatalogDropdown();
    expect(catalogDropdown).toMatchSnapshot();
  });

  it('should be opened by default if open is true', () => {
    const catalogDropdown = createCatalogDropdown({
      propsData: {
        open: true
      }
    });

    expect(catalogDropdown).toMatchSnapshot();
  });

  it('should be closed by default if open is false', () => {
    const catalogDropdown = createCatalogDropdown({
      propsData: {
        open: false
      }
    });

    expect(catalogDropdown).toMatchSnapshot();
  });

  it('should be disabled by default if disabled is true', () => {
    const catalogDropdown = createCatalogDropdown({
      propsData: {
        disabled: true
      }
    });

    expect(catalogDropdown).toMatchSnapshot();
  });

  it('should be enabled by default if disabled is false', () => {
    const catalogDropdown = createCatalogDropdown({
      propsData: {
        disabled: false
      }
    });

    expect(catalogDropdown).toMatchSnapshot();
  });
  
  describe('Items', () => {
    it('should show up to 8 items if the screen height by default', () => {
      const options = ['o_1', 'o_2', 'o_3', 'o_4', 'o_5', 'o_6', 'o_7', 'o_8', 'o_9'];
      
      window.innerHeight = 1200;

      const catalogDropdown = createCatalogDropdown({
        propsData: {
          options,
          open: true,
          disabled: false,
          limitHeight: false
        }
      });

      expect(catalogDropdown.vm.maxItemsScroll).toEqual(8);
    });

    it('should show up to 3 items if the screen height is "small"', () => {
      const options = ['o_1', 'o_2', 'o_3', 'o_4', 'o_5', 'o_6', 'o_7', 'o_8', 'o_9'];
      
      window.innerHeight = 300;

      const catalogDropdown = createCatalogDropdown({
        propsData: {
          options,
          open: true,
          disabled: false,
          limitHeight: false
        }
      });

      expect(catalogDropdown.vm.maxItemsScroll).toEqual(3);
    });

    it('should show up to 5 items if the screen height is "medium"', () => {
      const options = ['o_1', 'o_2', 'o_3', 'o_4', 'o_5', 'o_6', 'o_7', 'o_8', 'o_9'];
      
      window.innerHeight = 700;

      const catalogDropdown = createCatalogDropdown({
        propsData: {
          options,
          open: true,
          disabled: false,
          limitHeight: false
        }
      });

      expect(catalogDropdown.vm.maxItemsScroll).toEqual(5);
    });
  });

  describe('Search', () => {
    it('should be visible when there is an items overflow', () => {
      const options = ['o_1', 'o_2', 'o_3', 'o_4', 'o_5', 'o_6', 'o_7', 'o_8', 'o_9', 'o_10'];

      const catalogDropdown = createCatalogDropdown({
        propsData: {
          options,
          open: true,
          disabled: false,
          limitHeight: false
        }
      });

      expect(catalogDropdown).toMatchSnapshot();
    });

    it('should be hidden if there is no items overflow', () => {
      const options = ['o_1', 'o_2', 'o_3'];

      const catalogDropdown = createCatalogDropdown({
        propsData: {
          options,
          open: true,
          disabled: false,
          limitHeight: false
        }
      });

      expect(catalogDropdown).toMatchSnapshot();
    });
  });
});
