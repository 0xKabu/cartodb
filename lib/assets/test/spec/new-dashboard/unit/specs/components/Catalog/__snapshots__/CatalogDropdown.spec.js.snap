// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogDropdown.vue Search should be hidden if there is no items overflow 1`] = `
<div class="catalogDropdown is-open"><span class="title is-caption catalogDropdown__label"></span>
  <div class="catalogDropdown__container">
    <div class="text is-caption catalogDropdown__input">
      <catalogdropdownitem-stub option=""></catalogdropdownitem-stub>
    </div> <button class="catalogDropdown__close"><img src="../../assets/icons/common/dropdown-close.svg" width="16" height="20"></button>
    <!---->
    <ul class="catalogDropdown__list is-open ps-container">
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_1"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_2"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_3"></catalogdropdownitem-stub>
      </li>
    </ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue Search should be visible when there is an items overflow 1`] = `
<div class="catalogDropdown is-open"><span class="title is-caption catalogDropdown__label"></span>
  <div class="catalogDropdown__container">
    <div class="text is-caption catalogDropdown__input">
      <catalogdropdownitem-stub option=""></catalogdropdownitem-stub>
    </div> <button class="catalogDropdown__close"><img src="../../assets/icons/common/dropdown-close.svg" width="16" height="20"></button>
    <!---->
    <ul class="catalogDropdown__list is-open ps-container ps-active-y">
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_1"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_2"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_3"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_4"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_5"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_6"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_7"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_8"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_9"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_10"></catalogdropdownitem-stub>
      </li>
    </ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should be closed by default if open is false 1`] = `
<div class="catalogDropdown"><span class="title is-caption catalogDropdown__label"></span>
  <div class="catalogDropdown__container">
    <div class="text is-caption catalogDropdown__input">
      <catalogdropdownitem-stub option=""></catalogdropdownitem-stub>
    </div> <button class="catalogDropdown__close"><img src="../../assets/icons/common/dropdown-close.svg" width="16" height="20"></button>
    <!---->
    <ul class="catalogDropdown__list"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should be disabled by default if disabled is true 1`] = `
<div class="catalogDropdown is-disabled"><span class="title is-caption catalogDropdown__label"></span>
  <div class="catalogDropdown__container">
    <div class="text is-caption catalogDropdown__input">
      <catalogdropdownitem-stub option=""></catalogdropdownitem-stub>
    </div> <button class="catalogDropdown__close"><img src="../../assets/icons/common/dropdown-close.svg" width="16" height="20"></button>
    <!---->
    <ul class="catalogDropdown__list"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should be enabled by default if disabled is false 1`] = `
<div class="catalogDropdown"><span class="title is-caption catalogDropdown__label"></span>
  <div class="catalogDropdown__container">
    <div class="text is-caption catalogDropdown__input">
      <catalogdropdownitem-stub option=""></catalogdropdownitem-stub>
    </div> <button class="catalogDropdown__close"><img src="../../assets/icons/common/dropdown-close.svg" width="16" height="20"></button>
    <!---->
    <ul class="catalogDropdown__list"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should be opened by default if open is true 1`] = `
<div class="catalogDropdown is-open"><span class="title is-caption catalogDropdown__label"></span>
  <div class="catalogDropdown__container">
    <div class="text is-caption catalogDropdown__input">
      <catalogdropdownitem-stub option=""></catalogdropdownitem-stub>
    </div> <button class="catalogDropdown__close"><img src="../../assets/icons/common/dropdown-close.svg" width="16" height="20"></button>
    <!---->
    <ul class="catalogDropdown__list is-open ps-container"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should render correct contents 1`] = `
<div class="catalogDropdown is-open"><span class="title is-caption catalogDropdown__label">Title</span>
  <div class="catalogDropdown__container">
    <div class="text is-caption catalogDropdown__input">
      <catalogdropdownitem-stub option=""></catalogdropdownitem-stub>
    </div> <button class="catalogDropdown__close"><img src="../../assets/icons/common/dropdown-close.svg" width="16" height="20"></button>
    <!---->
    <ul class="catalogDropdown__list is-open ps-container"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;
